ompi/mca/coll/ucx/coll_ucx_module.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2011      Mellanox Technologies. All rights reserved.
+ * Copyright (c) 2014      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
+ * Copyright (c) 2019      Huawei Technologies Co., Ltd. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+#include "ompi/constants.h"
+#include "ompi/datatype/ompi_datatype.h"
+#include "ompi/mca/coll/base/coll_base_functions.h"
+#include "ompi/op/op.h"
+
+#include "coll_ucx.h"
+#include "coll_ucx_request.h"
+
+static int mca_coll_ucg_create(mca_coll_ucx_module_t *module,
+		struct ompi_communicator_t *comm)
+{
+    ucs_status_t error;
+    ucg_group_params_t args;
+    ucg_group_member_index_t rank_idx, my_idx;
+
+#if OMPI_GROUP_SPARSE
+    COLL_UCX_ERROR("Sparse process groups are not supported");
+    return UCS_ERR_UNSUPPORTED;
+#endif
+
+    /* Fill in group initialization parameters */
+    my_idx                 = ompi_comm_rank(comm);
+    args.member_count      = ompi_comm_size(comm);
+    args.cid               = ompi_comm_get_cid(comm);
+    /* args.mpi_copy_f     = ompi_datatype_copy_content_same_ddt; */
+    args.mpi_reduce_f      = ompi_op_reduce;
+    args.resolve_address_f = mca_coll_ucx_resolve_address;
+    args.release_address_f = mca_coll_ucx_release_address;
+    args.cb_group_obj      = comm;
+    args.distance          = alloca(args.member_count * sizeof(*args.distance));
+    if (args.distance == NULL) {
+        COLL_UCX_ERROR("Failed to allocate memory for %lu local ranks", args.member_count);
+        return OMPI_ERROR;
+
+    }
+
+    /* Generate (temporary) rank-distance array */
+    for (rank_idx = 0; rank_idx < args.member_count; rank_idx++) {
+    	struct ompi_proc_t *rank_iter =
+    			(struct ompi_proc_t*)ompi_comm_peer_lookup(comm, rank_idx);
+        rank_iter->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_COLL] = NULL;
+        if (rank_idx == my_idx) {
+            args.distance[rank_idx] = UCG_GROUP_MEMBER_DISTANCE_SELF;
+        } else if (OPAL_PROC_ON_LOCAL_SOCKET(rank_iter->super.proc_flags)) {
+            args.distance[rank_idx] = UCG_GROUP_MEMBER_DISTANCE_SOCKET;
+        } else if (OPAL_PROC_ON_LOCAL_HOST(rank_iter->super.proc_flags)) {
+            args.distance[rank_idx] = UCG_GROUP_MEMBER_DISTANCE_HOST;
+        } else {
+            args.distance[rank_idx] = UCG_GROUP_MEMBER_DISTANCE_NET;
+        }
+    }
+
+    /* TODO: use additional info, such as OMPI_COMM_IS_CART(comm) */
+    /* TODO: add support for comm->c_remote_comm  */
+    /* TODO: add support for sparse group storage */
+    error = ucg_group_create(mca_coll_ucx_component.ucg_worker, &args, &module->ucg_group);
+
+    /* Examine comm_new return value */
+    if (error != UCS_OK) {
```

Comment:
Ditto with above: this should be an OMPI_ error code, not UCS_ error code, right?  (I'm gussing UCS_OK == OMPI_SUCCESS, but still -- the abstraction layer should be preserved, right?